### DISCUSSION TITLE
defaults: use ripgrep (rg) for :grep if available

### DISCUSSION THREAD

**Discussion by LunarLambda** [neovim/neovim#28296] [Ideas]
I was playing with this idea the other day, and I would be happy to make a PR if people think this could be an accepted change.

For background, ripgrep is a faster grep written in Rust that's aimed at programmers and provides features like .gitignore support.
Many Neovim plugins like telescope.nvim and todo-comments.nvim already integrate with ripgrep out of the box for search support. Adopting this as a backwards-compatible alternative for :grep could be a great addition to Neovim's enhanced defaults.

I will probably publish this as a tiny plugin in the near future, but given how small and hands-off the entire thing is, I'm hopeful it could be considered as an inclusion into neovim core. So feedback about which functionality and default flags should be included are welcome.

At base, it's as simple as 

```lua
if vim.fn.executable('rg') == 1 then
    -- or 'rg --vimgrep $* /dev/null' to mirror default :grep behaviour
    -- -u to not consider .gitignore files, -uu to include dotfiles too
    vim.o.grepprg = 'rg --vimgrep '
    vim.o.grepformat = '%f:%l:%c:%m'
end
```

Additionally, ripgrep supports a vim-style 'smartcase' flag, so searching could be adapted to vim's settings automatically:

```lua
local function set_grepprg()
    local cmd = 'rg --vimgrep '
    if vim.o.ignorecase then
        if vim.o.smartcase then
            cmd = cmd .. '-S ' --smartcase
        else
            cmd = cmd .. '-i ' --ignore-case
        end
    end

    vim.o.grepprg = cmd
end

set_grepprg()
vim.o.grepformat = '%f:%l:%c:%m'

vim.api.nvim_create_autocmd('OptionSet', {
    group = vim.api.nvim_create_augroup('rg', { clear = true }),
    pattern = 'ignorecase,smartcase',
    callback = set_grepprg,
})
```

~~If accepted, Neovim could even bundle ripgrep by default on Windows in the future, since it is dual-licensed under MIT and the UNLICENSE.~~

EDIT: PR for the basic form of this is open https://github.com/neovim/neovim/pull/28324

**Comment by dundargoc** [comment]
> If accepted, Neovim could even bundle ripgrep by default on Windows in the future

Nah, very unlikely. We don't wanna add dependencies unless we absolutely have to and platform-specific solutions aren't desirable.

As for defaulting to ripgrepnif it exists: that could work, although unsure about what other people thinks of it. If the effort is small you could make a PR and let people say what they think of it. Proposals carry more weight if they come with an implementation.

**Reply by LunarLambda** [reply]
The code included here is basically the entire implementation. I can make a PR this weekend, no issue

**Reply by clason** [reply]
I think it's worth taking the temperature on this. Important considerations (which you should argue for):
* zero configuration (and minimal documentation)
* low to zero maintenance

If it's really that small, `_defaults.lua` may be the correct place for it (and phrasing this as "better default/out-of-the-box experience" will help with adoption). Should have a `checkhealth` test, though.

In general, we _are_ slowly opening up to more external and optional dependencies (`git`, `curl`, `tree-sitter`, ...) so this would fit into that strategy. The key point is "if `rg` is installed and found, that's a bonus; otherwise everything works as before". What we _don't_ want is getting issues about "my Neovim is not using `rg`, fix bug! [fifteen comments later] I installed it via NixOS btw. and it's in a different environment". The challenge will be to make that very clear from the beginning.

**Reply by LunarLambda** [reply]
I will keep these points in mind! I did think _defaults.lua would be an appropriate place. The only thing I'm not sure on is the autocmd. I think it'd be cool to have the :grep behaviour automatically line up with the case settings, but maybe it's overkill for now.

**Reply by clason** [reply]
Yeah, let's keep the autocommand out of it for now; those should be used sparingly because every new autocommand risks more and more unwanted interactions with other ones. We have to document this somewhere (e.g., `'grepprg'` as the new default) where we could add a -- shortened -- snippet for users to add to their `init.lua` if they want this.

**Reply by LunarLambda** [reply]
Got it, thank you for the advice.

**Comment by justinmk** [comment]
> > Neovim could even bundle ripgrep by default on Windows in the future
> 
> Nah, very unlikely. We don't wanna add dependencies unless we absolutely have to and platform-specific solutions aren't desirable.

Right. But we have other mechanisms to leverage, e.g. `:checkhealth` "Performance" section could show a hint recommending `rg`.

**Reply by LunarLambda** [reply]
There's a PR open right now that includes a basic healthcheck for it. I can add onto it if this is wanted.
